### PR TITLE
Experiments with non-serialised units

### PR DESCRIPTION
**Pull request by kwxm**
This is (the beginning of) an attempt at #1818. The basic idea is simple to implement and works well, but then I tried to be clever and use `coerce` and ran into problems: see the comments at the bottom of CBOR.hs.  The basic problem is that it's easy to make a non-serialised unit type which is coercible to and from (), but the coercion doesn't extend to full ASTs.  I spent quite a while trying to work out what was going on, but couldn't get to the bottom of it.  Comments on this would be appreciated.

This isn't a real PR: there's still more to do elsewhere, but I'll do that once I know whether there's any point in using `coerce`.  That can easily be abandoned in favour of `<$` if there's no way to convince the compiler to do what's required.

### CODE REVIEW

File: language-plutus-core/src/Language/PlutusCore/CBOR.hs [hunk 1]
```diff
@@ -272,3 +273,90 @@ instance Serialise UnknownDynamicBuiltinNameError
...
+   takes care of this for us: `serialise (InvisibleUnit () <$ prog)`
+    will serialise a program omitting the unit annotations, and
+   `deserialise` will give us back an `InvisibleUnit`-annotated program
+   which we can convert back to a ()-annotated program with `() <$ ...`.
+   The functions `serialiseProg`, `deserialiseProg` and `deserialiseProgOrFail`
+   do the conversions for us, so that the `InvisibleUnit` type doesn't have
+   to be exposed to uesrs.  These functions should always be used to serialise
+   and deserialise ()-annotated programs.
+
+   A possible problem is that `() <$ prog` has to visit every node in
```
Comment (reactormonk) [inline language-plutus-core/src/Language/PlutusCore/CBOR.hs:294]: You could also run `withMemory` over it and add the memory consumption annotation here. It's gotta happen somewhere.

Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/CBOR.hs:294]: Well, that also takes space. The point here is to make it small for transmission.

File: language-plutus-core/src/Language/PlutusCore/CBOR.hs [hunk 2]
```diff
@@ -272,3 +273,90 @@ instance Serialise UnknownDynamicBuiltinNameError
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (InternalTypeError uni ann)
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (TypeError uni ann)
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (Error uni ann)
+
+
+{- Serialising the unit annotation takes up space: () is converted to
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/CBOR.hs:278]: Make this a proper Note?

Nit: for long multiline doc comments I think we're mostly adding a line break and then not indenting the main body of the comment: makes editing easier.

**Comment by michaelpj** [comment]
The issue is this:
```
type role Type representational nominal representational
data Type tyname uni ann = ...
```
gives
```
    • Role mismatch on variable ann:
        Annotation says representational but role nominal is required
    • while checking a role annotation for ‘Type’
```
i.e. `Type` (and all the other stuff from there on up) is not representational in `ann`, which blocks coercing on that type.

Why is it not representational? Because it uses `tyname ann`. We can boil this down:
```
type role HKD representational representational
data HKD (f :: * -> *) (a :: *) = App (f a)
```
    • Role mismatch on variable a:
        Annotation says representational but role nominal is required
    • while checking a role annotation for ‘HKD’
```

This does, in fact, make sense: you could instantiate `HKD` with a `f` which is *not* representational in its argument, and then we shouldn't be able to coerce on `ann`.

You might think that GHC would be clever enough to nonetheless realise that things are okay if we use an `f` that's okay (like `Name`), but apparently it isn't.

What to do? One option is "nothing" and just live with it. I'm not sure we're really ready to stop being able to parameterize our name type, but I *do* think that we could drop annotations on names, since they're typically subsumed by the actual piece of syntax that contains the name. I'd be interested to see how this pans out.

**Comment by effectfully** [comment]
> You might think that GHC would be clever enough to nonetheless realise that things are okay if we use an `f` that's okay (like `Name`), but apparently it isn't.

Usually you can add a `Coercible (f a) (f ())` constraint and propagate it up until `f` gets instantiated to something that obviously satisfies the constraint.

**Comment by michaelpj** [comment]
> Usually you can add a Coercible (f a) (f ()) constraint and propagate it up until f gets instantiated to something that obviously satisfies the constraint.

But this would be in the `Coercible` instance for, say, `Type` itself. And I don't think you're supposed to write `Coercible` instances by hand, right?

**Comment by effectfully** [comment]
> But this would be in the `Coercible` instance for, say, `Type` itself.

Oh, I didn't realize that even

```haskell
newtype Unit = Unit ()

coerceType :: Type TyName uni Unit -> Type TyName uni ()
coerceType = coerce
```

wasn't working. I thought they were serious about creating instances on the fly:

> `Coercible` is a two-parameter class that has instances for types `a` and `b` if the compiler can infer that they have the same representation. This class does not have regular instances; instead they are created on-the-fly during type-checking.

Sorry for the noise then.

I vote for dropping annotations on names. They were required previously when we had names storing types, but all of that was removed and I don't think we now need them. I dropped annotations from constants when implementing extensible builtins, does anybody miss them?

**Comment by michaelpj** [comment]
Kenneth, do you want to try out removing annotations from names? Otherwise we can merge this as is and do it as an improvement.

File: language-plutus-core/src/Language/PlutusCore/CBOR.hs (hunk 1 above)
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/CBOR.hs:294]: On the deserilization side only.

**Comment by kwxm** [comment]
> The issue is this:
> 
> ```
> type role Type representational nominal representational
> data Type tyname uni ann = ...
> ```
> 
> gives
> 
> ```
>     • Role mismatch on variable ann:
>         Annotation says representational but role nominal is required
>     • while checking a role annotation for ‘Type’
> ```

Ok, thanks.  I did try adding roles to the `Term` type but I gave up because (a) I wasn't entirely sure what I was doing and (b) it looked as if I'd have to start adding them in lots of other places as well.

**Comment by kwxm** [comment]
> Kenneth, do you want to try out removing annotations from names? Otherwise we can merge this as is and do it as an improvement.

OK, I'll do that.  I imagine it'll be straightforward but tedious.

File: language-plutus-core/src/Language/PlutusCore/CBOR.hs [hunk 3]
```diff
@@ -272,3 +273,90 @@ instance Serialise UnknownDynamicBuiltinNameError
...
+   representation of the InvisibleUnit-annotated version.  Universes
+   seem to complicate matters, but I think there's more to it than
+   that because it doesn't work even if you fix uni to be
+   DefaultUniverse.
+-}
+
+deserialiseProg :: (Closed uni, uni `Everywhere` Serialise) =>
+                   BSL.ByteString -> Program TyName Name uni ()
+deserialiseProg s =
+    let p = deserialise s :: (Closed uni, uni `Everywhere` Serialise) =>
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/CBOR.hs:339]: You could avoid writing the `(Closed uni, uni `Everywhere` Serialise)` constraint here by binding `uni` explicitly in `deserialiseProg` and enabling `ScopedTypeVariables`, if it's not enabled already.

Comment (kwxm) [reply language-plutus-core/src/Language/PlutusCore/CBOR.hs]: I think that works now. I added a synonym for `(Closed uni, uni Everywhere Serialise)` just to make the signatures a bit more compact.

**Comment by kwxm** [comment]
Removing the name annotations worked!  `coerce` now works on entire ASTs, which was the point of the whole thing.  Thanks for the helpful comments.

**Comment by kwxm** [comment]
I've added functions `serialiseProg`, `deserialiseProg`, and `deserialiseProgOrFail` which serialise and deserialise ignoring units.  I'm now wondering whether I should use a typeclass for that .  For example in Ledger.Scripts we have
```
newtype Script = Script { unScript :: PLC.Program PLC.TyName PLC.Name PLC.DefaultUni () }
  deriving stock Generic
  deriving newtype (Serialise)
```
which will serialise things with units included.  It should be easy for people to use the new functions (and ideally we'd like to prevent them from using `Serialise` themselves, but I don't think that's possible when typeclasses are involved).  Maybe a new typeclass would help with that.

**Comment by michaelpj** [comment]
I don't think we need to be *too* stringent about this. This is, after all, a pretty specialized usecase specifically for minimizing size. The normal `Serialise` instance is perfectly sensible.

So really the only thing we need to be careful about is specifically `Script`. I'd just give that a hand-written `Serialise` instance with a note.

**Comment by michaelpj** [comment]
Also, it looks like you've done some kind of strange rebase here...

**Comment by kwxm** [comment]
> Also, it looks like you've done some kind of strange rebase here...

I rebased on master after merging the branch that removed unit annotations.  There's actually not that much code in this branch, so I could abandon it and start a new one if you want the history to look a bit cleaner.

**Comment by kwxm** [comment]
> I don't think we need to be _too_ stringent about this. This is, after all, a pretty specialized usecase specifically for minimizing size. The normal `Serialise` instance is perfectly sensible.

I'm just worried that it'd be too easy to use the wrong version if you're not aware of the issue.  I've put a warning about that at the top of CBOR.hs though.

**Comment by michaelpj** [comment]
Well, whatever you've done, it's seriously confused Github, which thinks there are 16k lines of changes. You have a real talent for confusing Git :p

> I'm just worried that it'd be too easy to use the wrong version if you're not aware of the issue.

For *who* to use the wrong version? There is precisely one place where we want to use this (serializing on-chain scripts in the ledger), and we need to be aware there, but everywhere else it really doesn't matter.

**Comment by kwxm** [comment]
> Well, whatever you've done, it's seriously confused Github, which thinks there are 16k lines of changes. You have a real talent for confusing Git :p

That's good: it makes it look as if I've been working hard.  I'll abandon ship and start again though.


> > I'm just worried that it'd be too easy to use the wrong version if you're not aware of the issue.
> 
> For _who_ to use the wrong version? There is precisely one place where we want to use this (serializing on-chain scripts in the ledger), and we need to be aware there, but everywhere else it really doesn't matter.

Well the serialised unit-less verison is the definitive one from the point of view of hashes for example, and you might well want to do things with them off-chain.  Admittedly you'd find out pretty quickly if you used the wrong thing to deserialise them though.  Also, Jann uses `encode` from the Serialase class a number of times in `Scripts.hs` to calculate hashes, and that'll have to be changed.

**Comment by kwxm** [comment]
I'm abandoning this branch because of the Git confusion.  New PR in #2005.

**Pull request closed without merging**